### PR TITLE
Make our compilation O(extensions) instead of O(extensions**2)

### PR DESCRIPTION
**Pull request by cramforce**
Closure compiler parses all files in the glob for module ids and dependencies only to drop 99% on a typical compilation.

This change make extensions only glob themselves and not other extensions and additonally makes the glob area sparser (e.g. only `integration.js` globs the 3p code like it should be).

Significant additional work can be done to further reduce the files included in a given compilation, but this should be a good first step.

### CODE REVIEW

**Comment by cramforce** [comment]
Makes our build about 1 minute faster. Better than nothing.

File: build-system/tasks/compile.js [hunk 1]
```diff
@@ -163,7 +163,7 @@ function compile(entryModuleFilenames, outputDir,
     // Instead of globbing all extensions, this will only add the actual
     // extension path for much quicker build times.
     entryModuleFilenames.forEach(function(filename) {
-      if (!filename.startsWith('extensions/')) {
+      if (!filename.indexOf('extensions/') == -1) {
```
Comment (jridgewell) [inline build-system/tasks/compile.js]: Gotta remove the `!`, too.

**Comment by cramforce** [comment]
I think the next big win would be to significantly reduce what we glob for extensions. Many need almost nothing.

**Comment by jridgewell** [comment]
> significantly reduce what we glob for extensions

Shouldn't we just be globbing for our entry points? So, `amp.js`, `amp-shadow.js`, `extension/([\w-]+)/*/\1.js`.

**Comment by cramforce** [comment]
@jridgewell unfortunately it doesn't walk from the entry points for the dep tree. You have to explicitly provide globs for each JS file that might be in the compilation unit.

Still, we can do much better, of course. If we had a dep tree from browserify, we could serialize a sparse glob for each entry point. Sounds like a neat weekend project :)

**Comment by erwinmombay** [comment]
@cramforce should be really easy to do. we basically already do that with dep-check. maybe do something like this on travis so we can just reuse the existing dep-check run result.

```bash
export AMP_SOURCE_GLOB=$(gulp dep-check)
gulp dist --srcs "${AMP_SOURCE_GLOB}"

```

or just expose a helper function on dep-check/index.js module that can be reused in compile.js
wrapping up some stuff and will do this.

**Comment by cramforce** [comment]
Yeah, it is a bit more complicated in practice (e.g. because of the dark
magic here with "fake-module" and "patched-module", but I think gulp
dep-deck could generate a JSON file that contains a list of deps for each
file.

From that we'd generate the srcs for closure by adding a few special cases
and we are good :)

**Comment by erwinmombay** [comment]
@cramforce i understand fake-module, but this is the first time I've head of patched module. Wouldn't it just a be a direct mapping of `some/path` to `fake-module/some/path` for specific files?

**Comment by cramforce** [comment]
Yes, in any case, this is such a small number of files that we can just
always include these paths in the glob.

**Comment by cramforce** [comment]
Should we get this in or you want to implement a permanent solution?

**Comment by erwinmombay** [comment]
@cramforce LGTM, sorry got to it late.

**Pull request merged**